### PR TITLE
Add support for streaming logs

### PR DESCRIPTION
**Pull request by andriisoldatenko**
Fix https://github.com/astronomer/astro-cli/issues/191

### CODE REVIEW

File: cmd/logs.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+// logsCmd represents the logs command
+var (
+	search string
+	follow bool
+
+	logsCmd = &cobra.Command{
+		Use:     "logs",
+		Aliases: []string{"log", "l"},
+		Short:   "show logs",
```
Comment (schnie) [inline cmd/logs.go]: "Stream logs from an Airflow deployment"

File: cmd/logs.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		Long: `A longer description that spans multiple lines and likely contains examples
```
Comment (schnie) [inline cmd/logs.go]: Same as short description but maybe include an example or two?

File: cmd/logs.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+and usage of using your command. For example:
+
+astro logs webserver`,
+	}
+
+	webserverLogsCmd = &cobra.Command{
+		Use:     "webserver",
+		Aliases: []string{"web", "w"},
+		Short:   "airflow webserver logs",
```
Comment (schnie) [inline cmd/logs.go]: "Stream logs from an Airflow webserver"

File: cmd/logs.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		Long:    "airflow webserver logs",
+		Args:    cobra.ExactArgs(1),
+		RunE:    webserverRemoteLogs,
+	}
+
+	schedulerLogsCmd = &cobra.Command{
+		Use:     "scheduler",
+		Aliases: []string{"sch", "s"},
+		Short:   "airflow scheduler logs",
```
Comment (schnie) [inline cmd/logs.go]: "Stream logs from an Airflow scheduler"

File: cmd/logs.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		Long:    "airflow scheduler logs",
+		Args:    cobra.ExactArgs(1),
+		RunE:    schedulerRemoteLogs,
+	}
+
+	workersLogsCmd = &cobra.Command{
+		Use:     "workers",
+		Aliases: []string{"workers", "worker", "wrk"},
+		Short:   "airflow workers logs",
```
Comment (schnie) [inline cmd/logs.go]: "Stream logs from Airflow workers"

File: cmd/logs.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		Long:    "airflow workers logs",
+		Args:    cobra.ExactArgs(1),
+		RunE:    workersRemoteLogs,
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(logsCmd)
+	webserverLogsCmd.Flags().StringVarP(&search, "search", "s", "", "Search term inside logs")
```
Comment (schnie) [inline cmd/logs.go]: Small, but one flag ends with a period and one doesn't. I don't have much preference, other than we match everything else :)

File: houston/websocket.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	defer ws.Close()
+
+	initSubscription := InitSubscription{Type: "connection_init", Payload: AuthPayload{Authorization: jwtToken}}
+	js, _ := json.Marshal(&initSubscription)
+	fmt.Printf("%s", js)
+
+	log.Printf("query: %s", initSubscription)
+	err = ws.WriteMessage(websocket.TextMessage, []byte(js))
+	if err != nil {
+		log.Printf("could not init connection: %s", err.Error())
```
Comment (schnie) [inline houston/websocket.go]: Capital C here?

File: houston/websocket.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	log.Printf("query: %v", queryMessage)
+	err = ws.WriteMessage(websocket.TextMessage, []byte(queryMessage))
+	if err != nil {
+		log.Printf("could not subscribe to logs: %s", err.Error())
```
Comment (schnie) [inline houston/websocket.go]: Capital C here too?

File: houston/websocket.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	log.Printf("listening for websocket events")
```
Comment (schnie) [inline houston/websocket.go]: Capital L

File: houston/websocket.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+			log.Printf("recv: %s", string(message))
+		}
+	}()
+
+	for {
+		select {
+		case <-done:
+			return nil
+		case <-interrupt:
+			log.Println("bye bye ... 👋")
```
Comment (schnie) [inline houston/websocket.go]: Capitalize first B?

**Comment by andriisoldatenko** [comment]
FYI
```
astro logs
Stream logs from an Airflow deployment. For example:

astro logs webserver YOU_DEPLOYMENT_ID -s string-to-find -f

Usage:
  astro logs [command]

Aliases:
  logs, log, l

Available Commands:
  scheduler   Stream logs from an Airflow scheduler
  webserver   Stream logs from an Airflow webserver
  workers     Stream logs from Airflow workers

Flags:
  -h, --help   help for logs

Use "astro logs [command] --help" for more information about a command.
```

```
astro logs webserver --help
Stream logs from an Airflow webserver. For example:

astro logs webserver YOU_DEPLOYMENT_ID -s string-to-find

Usage:
  astro logs webserver [flags]

Aliases:
  webserver, web, w

Flags:
  -f, --follow          Subscribe to watch more logs
  -h, --help            help for webserver
  -s, --search string   Search term inside logs
```

**Comment by andriisoldatenko** [comment]
@schnie PTAL
```
astro logs scheduler cjt4h94o70f970f521eri2v7m -s executor -f
Starting subscribing for logs:
[2019-03-28 22:51:19,297] {jobs.py:1662} INFO - Heartbeating the executor
[2019-03-28 22:51:20,299] {jobs.py:1662} INFO - Heartbeating the executor
[2019-03-28 22:51:21,301] {jobs.py:1662} INFO - Heartbeating the executor
[2019-03-28 22:51:22,303] {jobs.py:1662} INFO - Heartbeating the executor
[2019-03-28 22:51:23,419] {jobs.py:1662} INFO - Heartbeating the executor
[2019-03-28 22:51:24,420] {jobs.py:1662} INFO - Heartbeating the executor
[2019-03-28 22:51:25,420] {jobs.py:1662} INFO - Heartbeating the executor
^C2019/03/29 00:51:32 Bye bye ... 👋
```

**Comment by schnie** [comment]
@andriisoldatenko just took this for a test drive, looks awesome :)

Just a few more cosmetic things:

- "Starting subscribing for logs" -> "Waiting for logs..." (to match what we have in orbit) 
- The help messages for log commands start with a lowercase `H`. Can we capitalize those?
- I don't have emoji font in my terminal and I'd guess some other folks might not either. I like the wave but maybe we should remove it for now.

**Comment by andriisoldatenko** [comment]
@schnie 
1) Done
2) not sure:
```
// InitDefaultHelpFlag adds default help flag to c.
// It is called automatically by executing the c or by calling help and usage.
// If c already has help flag, it will do nothing.
func (c *Command) InitDefaultHelpFlag() {
	c.mergePersistentFlags()
	if c.Flags().Lookup("help") == nil {
     -- >>> usage := "help for "
		if c.Name() == "" {
			usage += "this command"
		} else {
			usage += c.Name()
		}
		c.Flags().BoolP("help", "h", false, usage)
	}
}
```
3) done

**Comment by schnie** [comment]
@andriisoldatenko looking good :+1: 

I just tried again locally and got a silent fail since my token had expired. Can we display the same "expired token" message for these commands if its expired?

**Comment by benjamingregory** [comment]
@andriisoldatenko -- Does this encompass logs for local deployments?

Right now you have to `docker logs -f XXXX` which isn't a great experience.

**Comment by andriisoldatenko** [comment]
@benjamingregory no. May be we need seperate ticket?
cc @schnie

**Comment by vparekh94** [comment]
@benjamingregory what about `astro airflow logs  -s -f` ?

**Comment by benjamingregory** [comment]
Ah - nvrmind then. Looks like I missed that one. 👍

**Pull request merged**